vscode-lean4/src/infoview.ts

```diff
@@ -568,23 +578,29 @@ export class InfoProvider implements Disposable {
         if (kind === 'above') {
             // in this case, assume that we actually want to insert at the same
             // indentation level as the neighboring text
-            const prev_line = editor.document.lineAt(pos.line - 1);
-            const spaces = prev_line.firstNonWhitespaceCharacterIndex;
+            const lines = text.split('\n');
+            const last_line = lines.length - 1;
+            const length_of_last_line = lines[last_line].length;
+            const current_line = editor.document.lineAt(pos.line);
+            const spaces = current_line.firstNonWhitespaceCharacterIndex;
             const margin_str = [...Array(spaces).keys()].map(x => ' ').join('');
-
             let new_command = text.replace(/\n/g, '\n' + margin_str);
-            new_command = `\n${margin_str}${new_command}`;
+            new_command = `${margin_str}${new_command}\n`;
+            const insertPosition = current_line.range.start;
 
             await editor.edit((builder) => {
-                builder.insert(prev_line.range.end, new_command);
+                builder.insert(insertPosition, new_command);
             });
-            editor.selection = new Selection(pos.line, spaces, pos.line, spaces);
+            // select the whole insert so it is easy for user to delete it.
+            editor.selection = new Selection(pos.line, spaces, pos.line + last_line, spaces + length_of_last_line);
         } else {
             await editor.edit((builder) => {
                 if (pos) builder.insert(pos, text);
             });
             editor.selection = new Selection(pos, pos)
         }
+        // ensure the text document has the keyboard focus.
+        void window.showTextDocument(editor.document, { viewColumn: editor.viewColumn, preserveFocus: false });
```

Comment:
```suggestion
        await window.showTextDocument(editor.document, { viewColumn: editor.viewColumn });
```
According to the docs, `false` is the default.

I'm not sure if it's better to leave the focus in the infoview or to move it to the editor (potentially also switching to a different file!).  But I can't see any obvious problems with focusing the modified file.

Reply:
It can't have been the default because I see a very real change with this.  Previously the infoview remains the active window (it is has the white close box in the tab).  Whereas with my code the active window moves back to my `Main.lean` file making it clear I can just start typing again...

Reply:
Previously = without `showTextDocument`, right?  I'm only suggesting to drop the `preserveFocus` argument, if merely dropping this argument results in the behavior that you describe then the docs are wrong and feel free to ignore my comment.

Reply:
It seems to default to `preserveFocus: false` for me, but I cannot see a guarantee of this in the docs (they just say what happens on `true`). It might even differ across platforms? Anyway I see no harm in being explicit, but we *do* need the `await` instead of the `void`.

Reply:
oh, can you educate me on why `await` is needed?  There's another `void window.showTextDocument` on line 557 that will also need fixing if this is required...

Reply:
In this case I can't see things going terribly wrong without it, but it's good style IMO to use `await` instead of `void` when making `async` calls in `async` contexts since it guarantees a kind of atomicity, i.e. by the time `handleInsertText` finishes we are guaranteed that the document has been shown. As far as I know, `void` just launches the action and leaves it to finish whenever (including after `handleInsertText` has returned), which can sometimes end up with hard-to-reproduce bugs due to interleaved execution.

Reply:
That's convincing, I added 2 missing awaits.  I've been using a lot of void on async calls for "fire and forget" semantics we so probably need to review all of these for consistency now and in future PR's.  Thanks for the education.